proof/refine/ARM/Ipc_R.thy

```diff
@@ -4604,68 +4493,446 @@ lemma maybe_return_sc_weak_valid_sched_action:
+  assumes "cap_relation reply_cap replyCap"
+      and "is_reply_cap reply_cap \<or> (reply_cap = cap.NullCap)"
+  shows "corres (=) (invs and valid_ready_qs and valid_cap reply_cap) invs'
+     (receive_ipc_preamble reply_cap thread)
+     (receiveIPC_preamble replyCap thread)"
+  supply if_split [split del]
+  apply (insert assms)
+  apply (clarsimp simp: receive_ipc_preamble_def receiveIPC_preamble_def)
+  apply (case_tac reply_cap; simp add: is_reply_cap_def)
+  apply (rule stronger_corres_guard_imp)
+    apply (clarsimp simp: liftM_def)
+    apply (rule corres_split_deprecated[OF _ get_reply_corres])
+      apply (rule corres_split_deprecated[OF corres_return_eq_same, where r'=dc], simp)
+        apply (rule corres_when2)
+         apply (clarsimp simp: reply_relation_def)
+         apply (rename_tac r, case_tac "replyTCB r"; simp)
+        apply (rename_tac r r')
+        apply (subgoal_tac "replyTCB r' = reply_tcb r", simp)
+         apply (rule cancel_ipc_corres)
+        apply (clarsimp simp: reply_relation_def)
        apply wpsimp
       apply wpsimp
-     apply (rule threadget_corres, simp add: tcb_relation_def)
-    apply wpsimp
-   apply wpsimp
-  apply (subst tcb_ep_find_index.simps)
-  apply (subst tcbEPFindIndex.simps)
-  apply (rule corres_guard_imp)
-    apply (rule corres_split_eqr)
-       apply (rule corres_split_eqr)
-          apply (rule corres_if, simp)
-           apply (rule corres_if, simp)
-            apply (rule corres_trivial, simp)
-           apply simp
-          apply (rule corres_trivial, simp)
-         apply (rule threadget_corres, simp add: tcb_relation_def)
-        apply (wp thread_get_wp)
-       apply (wp threadGet_wp)
-      apply (rule threadget_corres, simp add: tcb_relation_def)
-     apply (wp thread_get_wp)
-    apply (wpsimp wp: threadGet_wp)
-   apply (fastforce simp: projectKO_eq projectKO_tcb obj_at'_def)+
+     apply (wpsimp wp: get_simple_ko_wp)
+    apply (wpsimp wp: getReply_wp)
+   apply (clarsimp simp: valid_cap_def)
+   apply (frule (1) valid_objs_ko_at[OF invs_valid_objs])
+   apply (clarsimp simp: valid_obj_def valid_reply_def)
+  apply (clarsimp simp: valid_cap_def)
+  apply (erule cross_relF[OF _ reply_at'_cross_rel])
+  apply fastforce
   done
 
-lemma tcb_ep_dequeue_corres:
-  "qs = qs' \<Longrightarrow> corres (=) \<top> \<top> (tcb_ep_dequeue t qs) (tcbEPDequeue t qs')"
-  by (clarsimp simp: tcb_ep_dequeue_def tcbEPDequeue_def)
+lemma ri_preamble_vbreply:
+  "receive_ipc_preamble_rv p pp s \<Longrightarrow> valid_bound_reply pp s"
+  by (case_tac pp; clarsimp elim!: reply_at_ppred_reply_at)
+
+lemma ri_preamble_not_in_sc:
+  "\<lbrakk>sym_refs (state_refs_of s);
+    valid_replies s;
+    receive_ipc_preamble_rv reply_cap replyOpt s\<rbrakk>
+   \<Longrightarrow> valid_bound_obj (\<lambda>a b. a \<notin> fst ` replies_with_sc b) replyOpt s"
+  apply (case_tac replyOpt; simp)
+  apply (erule (1) no_tcb_not_in_replies_with_sc, simp)
+  done
+
+lemma receiveIPC_corres_helper:
```

Comment:
I know that getting these `do` blocks in the right format can sometimes be annoying and that these helper lemmas are often nice to use in the middle of big proofs, but in this case, is this just applying `K_bind_def` or something? If so, can you use a `simply only: K_bind_def` in the proof below instead of this helper lemma?

Reply:
Can you say more? The intention here was to change move the `when` block inside the case statement. Is there a way to do that directly with K_bind?

Reply:
Ah I think I completely misread this at first. I don't think there's a good way around this, so happy for you to leave it as is